File: java/client/src/test/java/org/signal/libsignal/net/AuthDevicesServiceTest.kt
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import kotlinx.coroutines.test.runTest
import org.signal.libsignal.internal.NativeTestingNice
import org.signal.libsignal.internal.RemoveDeviceOut
import org.signal.libsignal.internal.SetDeviceNameOut
import org.signal.libsignal.net.assertNonSuccess
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs

class AuthDevicesServiceTest {
  @Test
  fun testSetDeviceName() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_SetDeviceNameTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthDevicesService,
        invoke = { chat, req ->
          chat.setDeviceName(
            deviceId = req.id,
            encryptedName = req.encryptedName,
          )
        },
        check = { expected, actual ->
          when (expected) {
            SetDeviceNameOut.Success -> assertIs<RequestResult.Success<Unit>>(actual)
            SetDeviceNameOut.DeviceNotFound -> actual.assertNonSuccess<_, _, DeviceIdNotFoundException>()
          }
        },
      )
    }

  @Test
  fun testSetPushTokenFcm() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_SetPushTokenFcmTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthDevicesService,
        invoke = { chat, req ->
          chat.setPushToken(fcmToken = req)
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }

  @Test
  fun testRemoveDevice() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_RemoveDeviceTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthDevicesService,
        invoke = { chat, req ->
          chat.removeDevice(deviceId = req.id)
        },
        check = { expected, actual ->
          when (expected) {
            RemoveDeviceOut.Success -> assertIs<RequestResult.Success<Unit>>(actual)
          }
        },
      )
    }

  @Test
  fun testGetDevices() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_GetDevicesTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthDevicesService,
        invoke = { chat, req ->
          chat.getDevices()
        },
        check = { expected, actual ->
          assertEquals(
            expected.devices.map(LinkedDevice::fromInternal),
            assertIs<RequestResult.Success<List<LinkedDevice>>>(actual).result,
          )
        },
      )
    }

  @Test
  fun testClearPushToken() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_ClearPushTokenTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthDevicesService,
        invoke = { chat, req ->
          chat.clearPushToken()
        },
        check = { expected, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }
}

File: node/ts/test/chat/AuthDevicesServiceTest.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';

import * as NativeNice from '../../NativeNice.js';
import * as util from '../util.js';
import { AuthDevicesService } from '../../net.js';
import { connectAuth, defineTestGrpcCases } from './ServiceTestUtils.js';
import { ErrorCode, LibSignalErrorBase } from '../../Errors.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('AuthDevicesService', () => {
  describe('setDeviceName', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_SetDeviceNameTests(),
      connectAuth<AuthDevicesService>,
      async (
        chat: AuthDevicesService,
        { id, encryptedName }: NativeNice.SetDeviceNameArgs,
        resp: NativeNice.SetDeviceNameOut
      ) => {
        const out = chat.setDeviceName({
          deviceId: id,
          encryptedName,
        });
        switch (resp) {
          case 'success':
            await out;
            break;
          case 'deviceNotFound':
            expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.DeviceIdNotFound,
              });
            break;
          default:
            resp satisfies never;
        }
      }
    );
  });

  describe('removeDevice', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_RemoveDeviceTests(),
      connectAuth<AuthDevicesService>,
      async (
        chat: AuthDevicesService,
        { id }: NativeNice.RemoveDeviceArgs,
        resp: NativeNice.RemoveDeviceOut
      ) => {
        const out = chat.removeDevice({ deviceId: id });
        switch (resp) {
          case 'success':
            await out;
            break;
          default:
            resp satisfies never;
        }
      }
    );
  });

  describe('getDevices', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_GetDevicesTests(),
      connectAuth<AuthDevicesService>,
      async (
        chat: AuthDevicesService,
        _args: void,
        resp: NativeNice.GetDevicesOut
      ) => {
        const out = await chat.getDevices();
        expect(out).to.deep.equal(resp.devices);
      }
    );
  });

  describe('clearPushToken', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_ClearPushTokenTests(),
      connectAuth<AuthDevicesService>,
      async (chat: AuthDevicesService, _args: void, _resp: void) => {
        await chat.clearPushToken();
      }
    );
  });
});
